### ISSUE TITLE
enable Dataset level sampling strategy

### ISSUE THREAD

**Issue by AliTajeldin** [TresAmigosSD/SMV#525]
Users would often need to work on sample data in their day-to-day work and only use the full data in production/test run.  This ticket is not for implementing the actual sampling, that is beyond what SMV should do.   However, SMV should allow module authors to easily define the sampling strategies used on a per module basis and allow users to define sampling strategies at the project level.  There are multiple components to this issue:
* Allow users to define sample strategies for a given project.  These are just arbitrary labels.
* Allow users to specify a sampling strategy to use for a given run.
* Allow module authors to post process the output of a module/dataset based on the current active strategy.
* Persisted intermediate results should be affected by the current active sample strategy.  That is, their hash values or full path should change based on current strategy.  Otherwise, users would have to remove all intermediate files when switching strategies.

### Specify project strategies
```
# conf/smv-app-conf.props
smv.sampleStrategies = full, pct10, pct1
```

### Affect dataset output based on sample strategy
Datasets that depend on the current active strategy should define the `sample()` method which will be provided with the `DataFrame` that is the result of `doRun` and the current active strategy.  The `sample()` method can then choose to modify the output based on the active strategy.
```
class SomeCsvFile(SmvPyCsvFile):
  def inputPath: ...

  def sample(df, strategy):
    if strategy == pct10:
        return df.sample(0.1)
    elif strategy == pct1:
        return df.sample(0.01)
    else:
        return df
```

### Running with a sample strategy
If the user does not specify a sample strategy in `smv-pyrun` then the first strategy is picked.  Otherwise the user can use the `--sample` flag to SMV to specify the strategy.
```bash
$ smv-pyrun --run-app --sample pct10
```

**Note:** with shell, we would need to add an optional parameter to the `df()` method to allow the user to specify the sample strategy to use for resolving the dataframe.

**Comment by AliTajeldin** [comment]
**TBD:** should this just be handled by a user config parameter (#422).  For example:

```
# user config file
smv.config.keys = sample
smv.config.sample = pct10
```

And in the code just add the check for the sample at the end of the normal `run()` method.  Capturing this as a separate entity within SMV allows us to add something like "listSampleStrategies" function in shell and picking a default strategy, etc.

**Comment by ninjapapa** [comment]
implemented in #422
